### PR TITLE
Add deprecation warning for keys in `qml.specs`

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
In this [PR](https://github.com/PennyLaneAI/pennylane/pull/4015) we modified `qml.specs` to use the new `Resources` class. There are multiple keys in the dictionary that `qml.specs` returns, which have the same information as the class instance. This PR adds a deprecation warning to those fields. 

**Description of the Change:**
- Add a `SpecsDict` class which inherits from `dict`
- Update `__getitem__` method to raise a warning when accessing any of the following fields:
  - `gate_types`, 
  - `num_operations`, 
  - `num_used_wires`, 
  - `depth`

### CODE REVIEW

**Comment by Jaybsoni** [comment]
[sc-37603]

**Comment by trbromley** [comment]
Thanks @Jaybsoni! Please could you also add an entry to the deprecations docs file and changelog?

**Comment by Jaybsoni** [comment]
> Thanks @Jaybsoni! Please could you also add an entry to the deprecations docs file and changelog?

Do we need a changelog entry?

File: pennylane/tape/qscript.py [hunk 1]
```diff
@@ -1331,6 +1334,21 @@ def from_queue(cls, queue):
         return cls(*process_queue(queue))
 
 
+class SpecsDict(dict):
+    """A dictionary to track and warn about deprecated keys"""
+
+    deprecated_keys = ("gate_types", "num_operations", "num_used_wires", "depth")
+    new_keys = ("gate_types", "num_gates", "num_wires", "depth")
```
Comment (dwierichs) [inline pennylane/tape/qscript.py:1341]: Could also make this a `dict` between deprecated keys and new keys, making the lookup simpler.
However, as this is a warning/exceptional evaluation path, I suppose it does not matter too much.

File: pennylane/tape/qscript.py [hunk 2]
```diff
@@ -1331,6 +1334,21 @@ def from_queue(cls, queue):
...
+
+    def __getitem__(self, item):
+        if item in self.deprecated_keys:
+            warnings.warn(
+                f"The {item} key is deprecated and will be removed in the next release. "
```
Comment (dwierichs) [inline pennylane/tape/qscript.py:1346]: Which release? I think this should state the number.

File: doc/development/deprecations.rst [hunk 3]
```diff
@@ -17,13 +29,13 @@ Pending deprecations
...
   - Still accessible in v0.28, v0.29
-  - Will be removed in v0.30
+  - Will be removed in v0.31
 
 * The ``grouping`` module is deprecated. The functionality has been moved and
   reorganized in the new ``pauli`` module under ``pauli/utils.py`` or ``pauli/grouping/``.
 
   - Still accessible in v0.27, v0.28, v0.29
-  - Will be removed in v0.30
+  - Will be removed in v0.31
```
Comment (dwierichs) [inline doc/development/deprecations.rst:38]: This is unrelated I suppose? Is this a surprise/did we miss to realize what we announced here? (same above).
Also, should the "still accessible" lines include 0.30, then?

**Review by dwierichs — COMMENTED** [review]
Looks good to me, happy to approve after answers to three small comments.

File: pennylane/tape/qscript.py (hunk 1 above)
Comment (Jaybsoni) [reply pennylane/tape/qscript.py:1341]: Yes it would make the lookup simple. Initially the message was only displaying the deprecated keys, I added the new keys after the fact and completely missed this potential optimization.

File: pennylane/tape/qscript.py (hunk 2 above)
Comment (Jaybsoni) [reply pennylane/tape/qscript.py:1346]: It will be removed in V0.32! I don't know if this is a standard practice? I feel like it might be safer to leave it out incase the deprecation gets postponed?

File: doc/development/deprecations.rst (hunk 3 above)
Comment (Jaybsoni) [reply doc/development/deprecations.rst:38]: I just noticed it while I was adding this to the Changelog. As of now it hasn't been removed 😢, since this is the last day of feature freeze I doubt it will get removed this time. I just updated it here, I can add v0.30. Really this should also be changed in the RC-branch 👍🏼

File: pennylane/tape/qscript.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/tape/qscript.py:1346]: Mh. Good point :sweat_smile: 
Looking through the codebase, you seem to be right and we don't give version numbers usually :)

File: pennylane/tape/qscript.py (hunk 1 above)
Comment (dwierichs) [reply pennylane/tape/qscript.py:1341]: It really works both ways, no worries to optimize...

File: doc/development/deprecations.rst (hunk 3 above)
Comment (dwierichs) [reply doc/development/deprecations.rst:38]: Are you taking care of that? :)

Comment (Jaybsoni) [reply doc/development/deprecations.rst:38]: Made a separate PR for this 👍🏼

File: pennylane/tape/qscript.py (hunk 1 above)
Comment (Jaybsoni) [reply pennylane/tape/qscript.py]: Added 👍🏼

File: doc/development/deprecations.rst [hunk 4]
```diff
@@ -6,6 +6,18 @@ Deprecations
 Pending deprecations
 --------------------
 
+* The ``qml.specs`` dictionary will no longer support direct key access to certain keys. Instead
```
Comment (soranjh) [inline doc/development/deprecations.rst:9]: What will happen when `qml.specs` is used? The user gets all the warnings?

File: pennylane/tape/qscript.py [hunk 5]
```diff
@@ -1182,11 +1183,13 @@ def specs(self):
...
 
-            self._specs = {
-                "resources": resources,
-                "gate_sizes": defaultdict(int),
-                "gate_types": defaultdict(int),
-            }
+            self._specs = SpecsDict(
+                {
+                    "resources": resources,
+                    "gate_sizes": defaultdict(int),
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/tape/qscript.py:1189]: Yes, so all of the attributes are the same, and we added one new attribute `Resources`. Anytime a user uses an attribute that is duplicated (eg. `num_used_wires`) they will get a warning telling them it is deprecated and to use the associated field in `Resources` instead.

File: doc/development/deprecations.rst (hunk 4 above)
Comment (Jaybsoni) [reply doc/development/deprecations.rst:9]: Only if they call specs for a specific attribute that is duplicated. So `qml.specs['num_trainable_params']` will not raise any warning but `qml.specs["depth"]` will raise a deprecation warning.

Comment (soranjh) [reply doc/development/deprecations.rst:9]: How about `qml.specs` without any attribute specification? e.g.,

`>>> qml.specs(circuit)(params)`

There will be no warning I guess? Maybe we should add a general warning?

**Review by soranjh — APPROVED** [review]
Looks good, thanks @Jaybsoni. I just left a suggestion on raising a general warning when `qml.specs ` is called, even without requesting any specific attribute.

**Review by dwierichs — COMMENTED** [review]
Thanks @Jaybsoni 
Looks good to me ~~, I'm not entirely sure about the naming of `gate_sizes`: Isn't it storing gate _counts_ rather than the size of any gates? ~~

Nevermind! Happy to approve :)

**Review by dwierichs — APPROVED** [review]
